File: funcscript/parser/syntax/FuncScriptParser.GetSimpleString.cs
using System.Text;
using funcscript.model;

namespace funcscript.core
{
    public partial class FuncScriptParser
    {
        record GetStringResult(string Str, ParseNode Node,int NextIndex);
        static GetStringResult GetSimpleString(ParseContext context, int index)
        {
            var res = GetSimpleString(context,  "\"", index);
            if (res.NextIndex > index)
                return res;
            return GetSimpleString(context,  "'", index);
        }

        static GetStringResult GetSimpleString(ParseContext context,  string delimator, int index)
        {
            ParseNode parseNode = null;
            String str = null;
            var i = GetLiteralMatch(context, index, delimator).NextIndex;
            if (i == index)
                return new GetStringResult(null,null,index);
            int i2;
            var sb = new StringBuilder();
            while (true)
            {
                i2 = GetLiteralMatch(context, i, @"\n").NextIndex;
                if (i2 > i)
                {
                    i = i2;
                    sb.Append('\n');
                    continue;
                }

                i2 = GetLiteralMatch(context, i, @"\t").NextIndex;
                if (i2 > i)
                {
                    i = i2;
                    sb.Append('\t');
                    continue;
                }

                i2 = GetLiteralMatch(context, i, @"\\").NextIndex;
                if (i2 > i)
                {
                    i = i2;
                    sb.Append('\\');
                    continue;
                }

                i2 = GetLiteralMatch(context, i, @"\u").NextIndex;
                if (i2 > i)
                {
                    if (i + 6 <= context.Expression.Length) // Checking if there is enough room for 4 hex digits
                    {
                        var unicodeStr = context.Expression.Substring(i + 2, 4);
                        if (int.TryParse(unicodeStr, System.Globalization.NumberStyles.HexNumber, null,
                                out int charValue))
                        {
                            sb.Append((char)charValue);
                            i += 6; // Move past the "\uXXXX"
                            continue;
                        }
                    }
                }

                i2 = GetLiteralMatch(context, i, $@"\{delimator}").NextIndex;
                if (i2 > i)
                {
                    sb.Append(delimator);
                    i = i2;
                    continue;
                }

                if (i >= context.Expression.Length || GetLiteralMatch(context, i, delimator).NextIndex > i)
                    break;
                sb.Append(context.Expression[i]);
                i++;
            }

            i2 = GetLiteralMatch(context, i, delimator).NextIndex;
            if (i2 == i)
            {
                context.Serrors.Add(new SyntaxErrorData(i, 0, $"'{delimator}' expected"));
                return new GetStringResult(null,null,index);
            }

            i = i2;
            str = sb.ToString();
            parseNode = new ParseNode(ParseNodeType.LiteralString, index, i - index);
            return new GetStringResult(str,parseNode,i);
        }
    }
}

File: ts/core/src/parser/syntax/FuncScriptParser.GetSimpleString.ts
import { ParseContext, ParseNode, ParseNodeType, ParseResult, SyntaxErrorData } from "../FuncScriptParser.Main";
import { GetLiteralMatch } from "./FuncScriptParser.GetLiteralMatch";

export interface GetSimpleStringResult extends ParseResult {
    Str: string | null;
    ParseNode: ParseNode | null;
    NextIndex: number;
}

export function GetSimpleString(context: ParseContext, index: number): GetSimpleStringResult {
    let res = GetSimpleStringInternal(context, "\"", index);
    if (res.NextIndex > index) return res;
    return GetSimpleStringInternal(context, "'", index);
}

function GetSimpleStringInternal(context: ParseContext, delimator: string, index: number): GetSimpleStringResult {
    let parseNode: ParseNode | null = null;
    let str: string | null = null;
    let i = GetLiteralMatch(context, index, delimator).NextIndex;
    if (i === index) return { Str: null, ParseNode: null, NextIndex: index };

    let i2: number;
    let sb: string[] = [];

    while (true) {
        i2 = GetLiteralMatch(context, i, "\\n").NextIndex;
        if (i2 > i) {
            i = i2;
            sb.push('\n');
            continue;
        }

        i2 = GetLiteralMatch(context, i, "\\t").NextIndex;
        if (i2 > i) {
            i = i2;
            sb.push('\t');
            continue;
        }

        i2 = GetLiteralMatch(context, i, "\\\\").NextIndex;
        if (i2 > i) {
            i = i2;
            sb.push('\\');
            continue;
        }

        i2 = GetLiteralMatch(context, i, "\\u").NextIndex;
        if (i2 > i) {
            if (i + 6 <= context.Expression.length) {
                const unicodeStr = context.Expression.substring(i + 2, i + 6);
                const charValue = parseInt(unicodeStr, 16);
                if (!isNaN(charValue)) {
                    sb.push(String.fromCharCode(charValue));
                    i += 6;
                    continue;
                }
            }
        }

        i2 = GetLiteralMatch(context, i, "\\" + delimator).NextIndex;
        if (i2 > i) {
            sb.push(delimator);
            i = i2;
            continue;
        }

        if (i >= context.Expression.length || GetLiteralMatch(context, i, delimator).NextIndex > i) break;
        sb.push(context.Expression[i]);
        i++;
    }

    i2 = GetLiteralMatch(context, i, delimator).NextIndex;
    if (i2 === i) {
        context.SyntaxErrors.push(new SyntaxErrorData(i, 0, `'${delimator}' expected`));
        return { Str: null, ParseNode: null, NextIndex: index };
    }

    i = i2;
    str = sb.join('');
    parseNode = new ParseNode(ParseNodeType.LiteralString, index, i - index);
    return { Str: str, ParseNode: parseNode, NextIndex: i };
}